src/common.mak

```diff
@@ -0,0 +1,204 @@
+	LIB_PHOBOS = $(PHOBOS_BUILD_DIR)/libphobos2.a
+	LIB_PHOBOSSO = $(PHOBOS_BUILD_DIR)/libphobos2.so
+endif
+
+# build with shared library support
+# (defaults to true on supported platforms, can be overridden w/ make SHARED=0)
+SHARED=$(if $(findstring $(OS),linux freebsd),1,)
+LINKDL=$(if $(findstring $(OS),linux),-L-ldl,)
+
+################################################################################
+# Set CC and DMD
+################################################################################
+
+ifeq ($(OS),win32wine)
+	CC = wine dmc.exe
+	DMD = wine dmd.exe
+	RUN = wine
+else
+	DMD = $(DMD_BUILD_DIR)/dmd
+	ifeq ($(OS),win32)
+		CC = dmc
+	else
+		CC = cc
+	endif
+	RUN =
+endif
+
+################################################################################
+# Commonly used binaries
+################################################################################
+
+DUB=dub
+
+################################################################################
+# Platform-specific variables
+################################################################################
+
+# Set DOTOBJ and DOTEXE
+ifeq (,$(findstring win,$(OS)))
+	DOTOBJ:=.o
+	DOTEXE:=
+	PATHSEP:=/
+else
+	DOTOBJ:=.obj
+	DOTEXE:=.exe
+	PATHSEP:=$(shell echo "\\")
+endif
+
+ifeq (osx,$(OS))
+	DOTDLL:=.dylib
+	DOTLIB:=.a
+else
+	DOTDLL:=.so
+	DOTLIB:=.a
+endif
+
+################################################################################
+# Phobos DFLAGS (for linking with the currently built Phobos)
+################################################################################
+
+PHOBOS_DFLAGS=-conf= $(MODEL_FLAG) -I$(DRUNTIME_IMPORT_DIR) -I$(PHOBOS_IMPORT_DIR) -L-L$(PHOBOS_BUILD_DIR) $(PIC)
+ifeq (1,$(SHARED))
+PHOBOS_DFLAGS+=-defaultlib=libphobos2.so -L-rpath=$(PHOBOS_BUILD_DIR)
+endif
+
+################################################################################
+# Default DUB FLAGS
+################################################################################
+
+DUBFLAGS = --arch=$(subst 32,x86,$(subst 64,x86_64,$(MODEL)))
+
+################################################################################
+# Automatically create dlang/tools repository if non-existent
+################################################################################
+
+_all: all
+
+$(TOOLS_DIR):
+	git clone --depth=1 ${GIT_HOME}/$(@F) $@
+
+$(TOOLS_DIR)/checkwhitespace.d: | $(TOOLS_DIR)
+
+
+################################################################################
+# Common build rules
+#
+# Use _build to avoid conflicts with their real rules
+################################################################################
+
+$(DMD)_build: $(DMD)
+	make -C $(DMD_DIR)/src -f posix.mak BUILD=$(BUILD) OS=$(OS) MODEL=$(MODEL)
+	touch $@
+
+$(LIB_DRUNTIME)_build: $(LIB_DRUNTIME)
+	make -C $(DRUNTIME_DIR) -f posix.mak BUILD=$(BUILD) OS=$(OS) MODEL=$(MODEL)
+	touch $@
+
+ifeq (,$(findstring win,$(OS)))
+$(LIB_DRUNTIMESO)_build: $(LIB_DRUNTIME)_build
+endif
```

Comment:
This is also an optional idea: instead of redefining how DMD and DRuntime should be built over and over again,
the predefined variables with a `_build` suffix could be used.